### PR TITLE
Adding support for API tokens.

### PR DESCRIPTION
**Pull request by adam-in-ict**
# Description

Thank you for contributing to a JupiterOne integration!

Please include a summary of the change and which issue is fixed. Please also
include relevant motivation and context. List any dependencies that are required
for this change.

## Summary

Orca announced in June that they were migrating from API keys to API tokens.  This PR adds support for either option for the Orca integration.  We decrypt the secret provided and treat it as a token if we are able to find a URL in it.

## Type of change

Please leave any irrelevant options unchecked.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to
      not work as expected)
- [X] This change requires a documentation update

### General Development Checklist:

- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

### Integration Development Checklist:

Please leave any irrelevant options unchecked.

- [ ] I have checked for additional permissions required to call any new API
      endpoints, and have documented any additional permissions in
      `jupiterone.md`, where necessary.
- [ ] My changes properly paginate the target service provider's API
- [ ] My changes properly handle rate limiting of the target service provider's
      API
- [ ] My new integration step is instrumented to execute in the correct order
      using `dependsOn`
- [ ] I have referred to the
      [JupiterOne data model](https://github.com/JupiterOne/data-model/tree/main/src/schemas)
      to ensure that any new entities/relationships, and relevant properties,
      match the recommended model for this class of data
- [ ] I have updated the `CHANGELOG.md` file to describe my changes
- [ ] When changes include modifications to existing graph data ingestion, I've
      reviewed all existing managed questions referencing the entities,
      relationships, and their property names, to ensure those questions still
      function with my changes.

### CODE REVIEW

File: src/client.ts [hunk 1]
```diff
@@ -67,10 +67,22 @@ async function streamArray<T>(
...
 
+  // We have to handle API Keys and API Tokens differently.  This boolean is used
+  // to track which we're dealing with during a given execution.
+  private useTokenAuthorization: boolean = false;
+
   constructor(
     readonly config: IntegrationConfig,
     readonly logger: IntegrationLogger,
-  ) {}
+  ) {
```
Comment (zemberdotnet) [inline src/client.ts:77]: What do you think about having two separate clients?
I had to add in another auth method to rumble and took the approach @VDubber recommended.

This project is in a little different state so I’m also fine if we want to leave the project this way.

Comment (zemberdotnet) [reply src/client.ts:77]: https://github.com/JupiterOne/graph-rumble/pull/68

File: docs/development.md [hunk 2]
```diff
@@ -19,10 +19,14 @@ don't bother documenting that here.
 1. Log in to [Orca dashboard](https://app.orcasecurity.io) and navigate to
    [integrations](https://app.orcasecurity.io/integrations).
 2. Click the **Manage Keys** button to create a new Orca API token.
-3. Click "Generate A New Key" and make sure to supply it as the ENV variable
-   (CLIENT_SECRET=[token you've just generated]).
-4. Next, you also need to supply the account email as the ENV variable
-   (CLIENT_EMAIL=[the email you've used for authenticating]).
+3. Click "Create API Token".
+4. In order to ingest all data, provide the Administrator role. If a lesser role
```
Comment (zemberdotnet) [inline docs/development.md:23]: Do we need to update jupiterone.md as well?

File: src/client.ts (hunk 1 above)
Comment (adam-in-ict) [reply src/client.ts:77]: I like the idea and I like how it turned out in graph-rumble.  The only sticking point I have with it is this really only affects the few areas concerned with authentication.  If the iterate functions weren't going to be identical I think I'd be more inclined to do so.

I'll kick around a few ideas and see if I can get something similar that works nicely with Orca over the weekend.  Thanks!

File: docs/development.md (hunk 2 above)
Comment (adam-in-ict) [reply docs/development.md:23]: Whoops.  Good catch.  Thanks!  Updated in latest commit.

File: src/client.ts (hunk 1 above)
Comment (adam-in-ict) [reply src/client.ts:77]: Ok.  I spent maybe 10 minutes looking into creating a child APIClientTokenAuth class that inherits from the original APIClient and only redefines the three or so functions that are different and I like it way more.  I'll get it cleaned up and pushed before Monday.  Thanks for getting me to take another look at it!

File: docs/jupiterone.md [hunk 3]
```diff
@@ -33,10 +33,14 @@ If you need help with this integration, please contact
 1. Log in to [Orca dashboard](https://app.orcasecurity.io) and navigate to
    [integrations](https://app.orcasecurity.io/integrations).
 2. Click the **Manage Keys** button to create a new Orca API token.
-3. Click "Generate A New Key" and make sure to supply it as the ENV variable
-   (CLIENT_SECRET=[token you've just generated]).
-4. Next, you also need to supply the account email as the ENV variable
-   (CLIENT_EMAIL=[the email you've used for authenticating]).
+3. Click "Create API Token".
+4. In order to ingest all data, provide the Administrator role. If a lesser role
+   is provided, the integration will attempt to run as many steps as it is able
```
Comment (VDubber) [inline docs/jupiterone.md:38]: Can we be more specific about which data will or will not be collected?

File: src/client.ts [hunk 4]
```diff
@@ -531,9 +531,142 @@ export class APIClient {
...
+   */
+  async request<T>(
+    endpoint: string,
+    method: 'GET' | 'POST' = 'GET',
+    body?: any,
+  ): Promise<T> {
+    const url = `${this.config.clientBaseUrl}/api${endpoint}`;
+
+    const makeAuthenticatedRequest = async (authRetryCount: number) => {
+      if (authRetryCount > 0) {
```
Comment (zemberdotnet) [inline src/client.ts:601]: I don't see `authRetryCount` incremented anywhere. Am I missing it?

File: src/client.ts [hunk 5]
```diff
@@ -531,9 +531,142 @@ export class APIClient {
...
+        Authorization: `Token ${this.config.clientSecret}`,
+      },
+      body: body ? JSON.stringify(body) : undefined,
+    });
+
+    return response;
+  }
+
+  /**
+   * Verifies authentication by making lightweight HEAD request to https://api.orcasecurity.io/api/auth/tokens.
```
Comment (VDubber) [inline src/client.ts:565]: HEAD or GET?

Comment (VDubber) [reply src/client.ts:565]: Looks like this jsdoc needs to be updated

File: src/client.ts [hunk 6]
```diff
@@ -531,9 +531,142 @@ export class APIClient {
...
+        cause: new Error('Provider authentication failed'),
+        endpoint: `${this.config.clientBaseUrl}/api`,
+        status: response.status,
+        statusText: response.statusText,
+      });
+    }
+  }
+
+  /**
+   * Makes a GET request to the provided relative endpoint at https://api.orcasecurity.io/api.
```
Comment (VDubber) [inline src/client.ts:586]: nit: update jsdoc. baseUrl is provided by config now. GET or POST

File: src/client.ts (hunk 4 above)
Comment (adam-in-ict) [reply src/client.ts:601]: Good call.  It's a leftover from the API key method.  I'll get rid of it.  Thanks!

File: src/client.ts (hunk 5 above)
Comment (adam-in-ict) [reply src/client.ts:565]: Missed updating the comment when I updated it for the API token method.  I'll get it updated.  Thanks!

File: docs/jupiterone.md [hunk 7]
```diff
@@ -33,10 +33,14 @@ If you need help with this integration, please contact
...
+   to.
+5. Supply the created token as the ENV variable (CLIENT_SECRET="token you've
```
Comment (VDubber) [inline docs/jupiterone.md:40]: Do these steps need to be updated for the customer audience?

Comment (adam-in-ict) [reply docs/jupiterone.md:40]: Good point.  I was just matching the info in there originally, but this is all geared towards the developer.  Let me get it updated.  Thanks!

**Review by VDubber — APPROVED** [review]

**Pull request merged**